File: desktop/stella-browser/src/user-browser.ts
/**
 * User Browser Detection & Lifecycle
 *
 * Detects installed Chromium-based browsers on the user's system
 * and provides graceful shutdown for relaunching with Playwright pipes.
 */

import { execSync, spawn } from 'child_process';
import { existsSync } from 'fs';
import * as path from 'path';
import * as os from 'os';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export interface DetectedBrowser {
  name: string;
  executablePath: string;
  profileDir: string;
  processName: string; // Process name for shutdown (e.g., "chrome.exe", "Google Chrome")
}

interface BrowserCandidate {
  name: string;
  processName: string;
  win?: { exe: string; profile: string };
  mac?: { app: string; exe: string; profile: string };
  linux?: { exe: string; profile: string };
}

const BROWSERS: BrowserCandidate[] = [
  {
    name: 'Chrome',
    processName: 'chrome',
    win: {
      exe: 'C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe',
      profile: path.join(process.env.LOCALAPPDATA || '', 'Google', 'Chrome', 'User Data'),
    },
    mac: {
      app: 'Google Chrome',
      exe: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
      profile: path.join(os.homedir(), 'Library', 'Application Support', 'Google Chrome'),
    },
    linux: {
      exe: '/usr/bin/google-chrome',
      profile: path.join(os.homedir(), '.config', 'google-chrome'),
    },
  },
  {
    name: 'Edge',
    processName: 'msedge',
    win: {
      exe: 'C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe',
      profile: path.join(process.env.LOCALAPPDATA || '', 'Microsoft', 'Edge', 'User Data'),
    },
    mac: {
      app: 'Microsoft Edge',
      exe: '/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge',
      profile: path.join(os.homedir(), 'Library', 'Application Support', 'Microsoft Edge'),
    },
    linux: {
      exe: '/usr/bin/microsoft-edge',
      profile: path.join(os.homedir(), '.config', 'microsoft-edge'),
    },
  },
  {
    name: 'Brave',
    processName: 'brave',
    win: {
      exe: 'C:\\Program Files\\BraveSoftware\\Brave-Browser\\Application\\brave.exe',
      profile: path.join(
        process.env.LOCALAPPDATA || '',
        'BraveSoftware',
        'Brave-Browser',
        'User Data'
      ),
    },
    mac: {
      app: 'Brave Browser',
      exe: '/Applications/Brave Browser.app/Contents/MacOS/Brave Browser',
      profile: path.join(
        os.homedir(),
        'Library',
        'Application Support',
        'BraveSoftware',
        'Brave-Browser'
      ),
    },
    linux: {
      exe: '/usr/bin/brave-browser',
      profile: path.join(os.homedir(), '.config', 'BraveSoftware', 'Brave-Browser'),
    },
  },
  {
    name: 'Vivaldi',
    processName: 'vivaldi',
    win: {
      exe: path.join(process.env.LOCALAPPDATA || '', 'Vivaldi', 'Application', 'vivaldi.exe'),
      profile: path.join(process.env.LOCALAPPDATA || '', 'Vivaldi', 'User Data'),
    },
    mac: {
      app: 'Vivaldi',
      exe: '/Applications/Vivaldi.app/Contents/MacOS/Vivaldi',
      profile: path.join(os.homedir(), 'Library', 'Application Support', 'Vivaldi'),
    },
    linux: {
      exe: '/usr/bin/vivaldi',
      profile: path.join(os.homedir(), '.config', 'vivaldi'),
    },
  },
  {
    name: 'Arc',
    processName: 'Arc',
    mac: {
      app: 'Arc',
      exe: '/Applications/Arc.app/Contents/MacOS/Arc',
      profile: path.join(os.homedir(), 'Library', 'Application Support', 'Arc', 'User Data'),
    },
  },
];

/**
 * Detect all installed Chromium-based browsers on this system.
 */
export function detectBrowsers(): DetectedBrowser[] {
  const platform = process.platform;
  const detected: DetectedBrowser[] = [];

  for (const browser of BROWSERS) {
    let exePath: string | undefined;
    let profileDir: string | undefined;

    if (platform === 'win32' && browser.win) {
      exePath = browser.win.exe;
      profileDir = browser.win.profile;
    } else if (platform === 'darwin' && browser.mac) {
      exePath = browser.mac.exe;
      profileDir = browser.mac.profile;
    } else if (platform === 'linux' && browser.linux) {
      exePath = browser.linux.exe;
      profileDir = browser.linux.profile;
    }

    if (exePath && profileDir && existsSync(exePath)) {
      detected.push({
        name: browser.name,
        executablePath: exePath,
        profileDir,
        processName: browser.processName,
      });
    }
  }

  return detected;
}

/**
 * Detect which Chromium browser is currently running, or fall back to first installed.
 */
export function detectDefaultBrowser(): DetectedBrowser | null {
  const installed = detectBrowsers();
  if (installed.length === 0) return null;

  // Check which one is currently running
  for (const browser of installed) {
    if (isBrowserRunning(browser)) {
      return browser;
    }
  }

  // None running — return first installed
  return installed[0];
}

/**
 * Check if a browser process is currently running.
 */
function isBrowserRunning(browser: DetectedBrowser): boolean {
  try {
    if (process.platform === 'win32') {
      const output = execSync(`tasklist /FI "IMAGENAME eq ${browser.processName}.exe" /NH`, {
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      });
      return output.includes(browser.processName);
    } else {
      // macOS / Linux: use pgrep
      execSync(`pgrep -x "${browser.processName}"`, {
        stdio: ['pipe', 'pipe', 'pipe'],
      });
      return true;
    }
  } catch {
    return false;
  }
}

/**
 * Gracefully shut down a running browser so we can relaunch it with Playwright pipes.
 * Waits up to `timeoutMs` for the process to exit.
 */
export async function gracefulShutdown(
  browser: DetectedBrowser,
  timeoutMs: number = 10000
): Promise<void> {
  if (!isBrowserRunning(browser)) return;

  console.log(`[UserBrowser] Shutting down ${browser.name}...`);

  if (process.platform === 'win32') {
    // Graceful close on Windows (no /F flag)
    try {
      execSync(`taskkill /IM ${browser.processName}.exe`, {
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      });
    } catch {
      // May fail if already closing
    }
  } else if (process.platform === 'darwin') {
    // macOS: use AppleScript for graceful quit
    const candidate = BROWSERS.find((b) => b.name === browser.name);
    const appName = candidate?.mac?.app;
    if (appName) {
      try {
        execSync(`osascript -e 'tell application "${appName}" to quit'`, {
          encoding: 'utf-8',
          stdio: ['pipe', 'pipe', 'pipe'],
        });
      } catch {
        // Fallback to SIGTERM
        try {
          execSync(`pkill -TERM "${browser.processName}"`, {
            stdio: ['pipe', 'pipe', 'pipe'],
          });
        } catch {
          // Already dead
        }
      }
    }
  } else {
    // Linux: SIGTERM
    try {
      execSync(`pkill -TERM "${browser.processName}"`, {
        stdio: ['pipe', 'pipe', 'pipe'],
      });
    } catch {
      // Already dead
    }
  }

  // Wait for the process to actually exit
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (!isBrowserRunning(browser)) {
      console.log(`[UserBrowser] ${browser.name} shut down successfully`);
      return;
    }
    await new Promise((r) => setTimeout(r, 500));
  }

  // If still running after timeout, force kill
  console.log(`[UserBrowser] ${browser.name} didn't shut down gracefully, force killing...`);
  try {
    if (process.platform === 'win32') {
      execSync(`taskkill /F /IM ${browser.processName}.exe`, {
        stdio: ['pipe', 'pipe', 'pipe'],
      });
    } else {
      execSync(`pkill -9 "${browser.processName}"`, {
        stdio: ['pipe', 'pipe', 'pipe'],
      });
    }
  } catch {
    // Best effort
  }

  // Wait a bit more for cleanup
  await new Promise((r) => setTimeout(r, 1000));
}

/**
 * Relaunch the user's browser with --silent-debugger-extension-api flag.
 * This suppresses the "started debugging" infobar when the extension uses chrome.debugger.
 * Returns the detected browser info.
 */
export async function relaunchForExtensionBridge(
  extraArgs: string[] = []
): Promise<DetectedBrowser> {
  const detected = detectDefaultBrowser();
  if (!detected) {
    throw new Error(
      'No Chromium browser found. Install Chrome, Edge, Brave, or another Chromium browser.'
    );
  }

  console.log(`[UserBrowser] Detected ${detected.name} at ${detected.executablePath}`);

  // Shut down existing browser
  await gracefulShutdown(detected);
  // Wait for profile lock release
  await new Promise((r) => setTimeout(r, 1500));

  // Relaunch with the silent debugger flag + load extension from source
  const extensionPath = path.resolve(__dirname, '..', 'extension');
  const args = [
    '--silent-debugger-extension-api',
    '--no-first-run',
    '--no-default-browser-check',
    '--hide-crash-restore-bubble',
    '--disable-session-crashed-bubble',
    '--restore-last-session',
    `--load-extension=${extensionPath}`,
    ...extraArgs,
  ];

  console.log(`[UserBrowser] Relaunching ${detected.name} with flags:`, args);
  const child = spawn(detected.executablePath, args, {
    stdio: 'ignore',
    detached: true,
  });
  child.unref();

  // Wait for the browser to start
  const start = Date.now();
  while (Date.now() - start < 15000) {
    if (isBrowserRunning(detected)) {
      console.log(`[UserBrowser] ${detected.name} is running`);
      return detected;
    }
    await new Promise((r) => setTimeout(r, 500));
  }

  throw new Error(`${detected.name} failed to start within 15 seconds`);
}

File: desktop/stella-browser/cli/src/native/user_browser.rs
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::thread;
use std::time::{Duration, Instant};

#[derive(Debug, Clone)]
pub struct DetectedBrowser {
    pub name: &'static str,
    pub executable_path: PathBuf,
    pub profile_dir: PathBuf,
    pub process_name: &'static str,
}

struct BrowserCandidate {
    name: &'static str,
    process_name: &'static str,
    win_exe: Option<&'static str>,
    win_profile: Option<Vec<String>>,
    mac_exe: Option<&'static str>,
    mac_profile: Option<Vec<String>>,
    linux_exe: Option<&'static str>,
    linux_profile: Option<Vec<String>>,
}

fn browser_candidates() -> Vec<BrowserCandidate> {
    vec![
        BrowserCandidate {
            name: "Chrome",
            process_name: "chrome",
            win_exe: Some(r"C:\Program Files\Google\Chrome\Application\chrome.exe"),
            win_profile: Some(vec!["Google".into(), "Chrome".into(), "User Data".into()]),
            mac_exe: Some("/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"),
            mac_profile: Some(vec![
                "Library".into(),
                "Application Support".into(),
                "Google Chrome".into(),
            ]),
            linux_exe: Some("/usr/bin/google-chrome"),
            linux_profile: Some(vec![".config".into(), "google-chrome".into()]),
        },
        BrowserCandidate {
            name: "Edge",
            process_name: "msedge",
            win_exe: Some(r"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"),
            win_profile: Some(vec!["Microsoft".into(), "Edge".into(), "User Data".into()]),
            mac_exe: Some("/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge"),
            mac_profile: Some(vec![
                "Library".into(),
                "Application Support".into(),
                "Microsoft Edge".into(),
            ]),
            linux_exe: Some("/usr/bin/microsoft-edge"),
            linux_profile: Some(vec![".config".into(), "microsoft-edge".into()]),
        },
        BrowserCandidate {
            name: "Brave",
            process_name: "brave",
            win_exe: Some(r"C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe"),
            win_profile: Some(vec![
                "BraveSoftware".into(),
                "Brave-Browser".into(),
                "User Data".into(),
            ]),
            mac_exe: Some("/Applications/Brave Browser.app/Contents/MacOS/Brave Browser"),
            mac_profile: Some(vec![
                "Library".into(),
                "Application Support".into(),
                "BraveSoftware".into(),
                "Brave-Browser".into(),
            ]),
            linux_exe: Some("/usr/bin/brave-browser"),
            linux_profile: Some(vec![
                ".config".into(),
                "BraveSoftware".into(),
                "Brave-Browser".into(),
            ]),
        },
        BrowserCandidate {
            name: "Vivaldi",
            process_name: "vivaldi",
            win_exe: None,
            win_profile: Some(vec!["Vivaldi".into(), "User Data".into()]),
            mac_exe: Some("/Applications/Vivaldi.app/Contents/MacOS/Vivaldi"),
            mac_profile: Some(vec![
                "Library".into(),
                "Application Support".into(),
                "Vivaldi".into(),
            ]),
            linux_exe: Some("/usr/bin/vivaldi"),
            linux_profile: Some(vec![".config".into(), "vivaldi".into()]),
        },
        BrowserCandidate {
            name: "Arc",
            process_name: "Arc",
            win_exe: None,
            win_profile: None,
            mac_exe: Some("/Applications/Arc.app/Contents/MacOS/Arc"),
            mac_profile: Some(vec![
                "Library".into(),
                "Application Support".into(),
                "Arc".into(),
                "User Data".into(),
            ]),
            linux_exe: None,
            linux_profile: None,
        },
    ]
}

pub fn detect_default_browser() -> Option<DetectedBrowser> {
    let installed = detect_browsers();
    if installed.is_empty() {
        return None;
    }

    for browser in &installed {
        if is_browser_running(browser) {
            return Some(browser.clone());
        }
    }

    installed.into_iter().next()
}

pub fn find_extension_dir() -> Option<PathBuf> {
    let mut candidates = Vec::new();

    if let Ok(home) = std::env::var("STELLA_BROWSER_HOME") {
        candidates.push(PathBuf::from(home).join("extension"));
    }

    if let Ok(exe) = std::env::current_exe() {
        if let Some(dir) = exe.parent() {
            candidates.push(dir.join("extension"));
            candidates.push(dir.join("../extension"));
            candidates.push(dir.join("../../extension"));
            candidates.push(dir.join("../../../extension"));
        }
    }

    if let Ok(cwd) = std::env::current_dir() {
        candidates.push(cwd.join("extension"));
        candidates.push(cwd.join("../extension"));
    }

    candidates
        .into_iter()
        .find(|path| path.exists() && path.is_dir())
        .and_then(|path| path.canonicalize().ok().or(Some(path)))
}

pub async fn relaunch_for_extension_bridge(extra_args: &[String]) -> Result<DetectedBrowser, String> {
    let extra_args = extra_args.to_vec();
    tokio::task::spawn_blocking(move || relaunch_for_extension_bridge_blocking(&extra_args))
        .await
        .map_err(|e| format!("Browser relaunch task failed: {}", e))?
}

fn relaunch_for_extension_bridge_blocking(extra_args: &[String]) -> Result<DetectedBrowser, String> {
    let detected = detect_default_browser().ok_or_else(|| {
        "No Chromium browser found. Install Chrome, Edge, Brave, or another Chromium browser."
            .to_string()
    })?;

    graceful_shutdown(&detected, Duration::from_secs(10));
    thread::sleep(Duration::from_millis(1500));

    let extension_path = find_extension_dir()
        .ok_or_else(|| "Could not find extension directory for user-browser relaunch".to_string())?;

    let mut args = vec![
        "--silent-debugger-extension-api".to_string(),
        "--no-first-run".to_string(),
        "--no-default-browser-check".to_string(),
        "--hide-crash-restore-bubble".to_string(),
        "--disable-session-crashed-bubble".to_string(),
        "--restore-last-session".to_string(),
        format!("--disable-extensions-except={}", extension_path.display()),
        format!("--load-extension={}", extension_path.display()),
    ];
    if extra_args.is_empty() {
        // Load one normal web page so the extension's content-script keepalive
        // can wake the MV3 service worker after a cold browser relaunch.
        args.push("https://example.com/".to_string());
    } else {
        args.extend(extra_args.iter().cloned());
    }

    spawn_detached(&detected.executable_path, &args)?;

    let deadline = Instant::now() + Duration::from_secs(15);
    while Instant::now() < deadline {
        if is_browser_running(&detected) {
            return Ok(detected);
        }
        thread::sleep(Duration::from_millis(500));
    }

    Err(format!("{} failed to start within 15 seconds", detected.name))
}

fn detect_browsers() -> Vec<DetectedBrowser> {
    browser_candidates()
        .into_iter()
        .filter_map(|candidate| detect_browser(candidate))
        .collect()
}

fn detect_browser(candidate: BrowserCandidate) -> Option<DetectedBrowser> {
    let (exe, profile) = if cfg!(target_os = "windows") {
        (
            candidate
                .win_exe
                .map(PathBuf::from)
                .or_else(|| {
                    let local_app_data = std::env::var("LOCALAPPDATA").ok()?;
                    Some(PathBuf::from(local_app_data).join("Vivaldi/Application/vivaldi.exe"))
                }),
            candidate
                .win_profile
                .and_then(build_local_app_data_path),
        )
    } else if cfg!(target_os = "macos") {
        (
            candidate.mac_exe.map(PathBuf::from),
            candidate.mac_profile.and_then(build_home_path),
        )
    } else if cfg!(target_os = "linux") {
        (
            candidate.linux_exe.map(PathBuf::from),
            candidate.linux_profile.and_then(build_home_path),
        )
    } else {
        (None, None)
    };

    let executable_path = exe?;
    let profile_dir = profile?;

    if executable_path.exists() {
        Some(DetectedBrowser {
            name: candidate.name,
            executable_path,
            profile_dir,
            process_name: candidate.process_name,
        })
    } else {
        None
    }
}

fn build_home_path(parts: Vec<String>) -> Option<PathBuf> {
    let mut path = dirs::home_dir()?;
    for part in parts {
        path.push(part);
    }
    Some(path)
}

fn build_local_app_data_path(parts: Vec<String>) -> Option<PathBuf> {
    let mut path = PathBuf::from(std::env::var("LOCALAPPDATA").ok()?);
    for part in parts {
        path.push(part);
    }
    Some(path)
}

fn is_browser_running(browser: &DetectedBrowser) -> bool {
    if cfg!(target_os = "windows") {
        Command::new("tasklist")
            .args([
                "/FI",
                &format!("IMAGENAME eq {}.exe", browser.process_name),
                "/NH",
            ])
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .output()
            .ok()
            .and_then(|output| String::from_utf8(output.stdout).ok())
            .is_some_and(|output| output.contains(browser.process_name))
    } else {
        Command::new("pgrep")
            .args(["-x", browser.process_name])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .map(|status| status.success())
            .unwrap_or(false)
    }
}

fn graceful_shutdown(browser: &DetectedBrowser, timeout: Duration) {
    if !is_browser_running(browser) {
        return;
    }

    if cfg!(target_os = "windows") {
        let _ = Command::new("taskkill")
            .args(["/IM", &format!("{}.exe", browser.process_name)])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
    } else if cfg!(target_os = "macos") {
        let app_name = browser.name;
        let apple_script = format!(r#"tell application "{}" to quit"#, app_name);
        let graceful = Command::new("osascript")
            .args(["-e", &apple_script])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .map(|status| status.success())
            .unwrap_or(false);

        if !graceful {
            let _ = Command::new("pkill")
                .args(["-TERM", browser.process_name])
                .stdout(Stdio::null())
                .stderr(Stdio::null())
                .status();
        }
    } else {
        let _ = Command::new("pkill")
            .args(["-TERM", browser.process_name])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
    }

    let deadline = Instant::now() + timeout;
    while Instant::now() < deadline {
        if !is_browser_running(browser) {
            return;
        }
        thread::sleep(Duration::from_millis(500));
    }

    if cfg!(target_os = "windows") {
        let _ = Command::new("taskkill")
            .args(["/F", "/IM", &format!("{}.exe", browser.process_name)])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
    } else {
        let _ = Command::new("pkill")
            .args(["-9", browser.process_name])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
    }

    thread::sleep(Duration::from_secs(1));
}

fn spawn_detached(executable_path: &Path, args: &[String]) -> Result<(), String> {
    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt;

        let mut command = Command::new(executable_path);
        command.args(args);
        unsafe {
            command.pre_exec(|| {
                libc::setsid();
                Ok(())
            });
        }
        command
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
            .map_err(|e| format!("Failed to relaunch browser: {}", e))?;
    }

    #[cfg(windows)]
    {
        use std::os::windows::process::CommandExt;

        const CREATE_NEW_PROCESS_GROUP: u32 = 0x00000200;
        const DETACHED_PROCESS: u32 = 0x00000008;

        Command::new(executable_path)
            .args(args)
            .creation_flags(CREATE_NEW_PROCESS_GROUP | DETACHED_PROCESS)
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
            .map_err(|e| format!("Failed to relaunch browser: {}", e))?;
    }

    Ok(())
}
